lib/attrsets.nix

```diff
@@ -823,6 +823,57 @@ rec {
         ) a (attrNames n)
     ) {} list_of_attrs;
 
+  /**
+    Flatten an attribute set to the bottom name and value pairs as
+    one singular final attribute set, stripping everything else
+    above.
+
+    In a few words, this does what `lib.lists.flatten` does, but for
+    attribute sets, rather than lists.
+
+
+    # Inputs
+
+    `attrs`
+
+    : The top-level attribute set to flatten.
+
+    # Type
+
+    ```
+    flattenAttrs :: AttrSet -> AttrSet
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.attrsets.flattenAttrs` usage example
+
+    ```nix
+    flattenAttrs { a.b.c.d = "e"; one.two.three.four.five = 6; foo.bar = null; }
+    => { d = "e"; five = 6; bar = null; }
+    ```
+
+    :::
+  */
+  flattenAttrs = attrs:
```

Comment:
I would want that function but with paths having the option of beeing preserved or an injectable function that maps them to the new name in the output.

Reply:
I have some idea of what you are describing. Could you provide some pseudo-example of what that would like look like, such as `<function> <what-args> -> <expected-result>`? Some brief example of what you "envision" would be helpful to know how to orient the reimplementation.

Reply:
This function reminds me of [`collect`](https://noogle.dev/f/lib/collect).

It iterates recursively over attribute values. If the predicate is true it adds the mapped value to `accumulator` (a list)

Its a mixture of `mapAttrsRecursiveCond` and `collect`

Example (lets call it `collect'`? )

```nix
/*
  Arguments:
  - cond: whether to recurse into the attribute set.
  - f: A function that produces the value
  - attrs: An attribute set
*/
#           ↓ cond             ↓ f
collect' (set: isAttrs set) (path: value: { name = head path; inherit value; } ) { a.foo = 1; b.foo = 2; }
> [ {name = "foo" value = 1; } { name = foo;  value = 2; } ]

# in your case you can create an attribute of this
lib.listToAttrs [ {name = "foo" value = 1; } { name = foo;  value = 2; }  ] 
> { foo = 1; }
# --- or
lib.zipAttrsWith (_: values: values ) ( map (elem: lib.listToAttrs [elem]) [ {name = "foo"; value = 1; } { name = "foo";  value = 2; }  ])
{
  foo = [
    1
    2
  ];
}
```


Maybe this function could possibly benefit the module system ? @roberth do you have any ideas or thoughts ?

Reply:
> If the predicate is true it adds the mapped value to `accumulator` (a list)

Seems that it could be optimized a bit then. Lists are quite inefficient because they don't have an efficient append.

`concatMapAttrs` + a recursive call is sufficient for a whole bunch of attribute set transformations. If you need context from the parent, you can pass it down as a parameter in the recursive call. No need for a list that way.